src/GafferScene/InteractiveRender.cpp

```diff
@@ -217,6 +286,437 @@ void InteractiveRender::parentChanged( GraphComponent *child, GraphComponent *ol
+		{
+			// close pending attribute blocks:
+			while( m_attrBlockCounter )
+			{
+				--m_attrBlockCounter;
+				m_renderer->attributeEnd();
+			}
+		}
+		
+		void *operator()( void *item )
+		{
+			SceneGraph *s = (SceneGraph*)item;
+			ScenePlug::ScenePath path;
+			s->path( path );
+			
+			std::string name;
+			ScenePlug::pathToString( path, name );
+
+			try
+			{
+				if( !m_editMode )
+				{
+					// outputting scene for the first time - do some attribute block tracking:
+					if( path.size() )
+					{
+						for( int i = m_previousPath.size(); i >= (int)path.size(); --i )
+						{
+							--m_attrBlockCounter;
+							m_renderer->attributeEnd();
+						}
+					}
+
+					m_previousPath = path;
+
+					++m_attrBlockCounter;
+					m_renderer->attributeBegin();
+
+					// set the name for this location:
+					m_renderer->setAttribute( "name", new StringData( name ) );
+
+				}
+
+				// transform:
+				if( s->m_transform && !m_editMode )
+				{
+					m_renderer->concatTransform( s->m_transform->readable() );
+					s->m_transform = 0;
+				}
+
+				// attributes:
+				if( s->m_attributes )
+				{
+					if( m_editMode )
+					{
+						CompoundDataMap parameters;
+						parameters["exactscopename"] = new StringData( name );
+						m_renderer->editBegin( "attribute", parameters );
+					}
+
+					for( CompoundObject::ObjectMap::const_iterator it = s->m_attributes->members().begin(), eIt = s->m_attributes->members().end(); it != eIt; it++ )
+					{
+						if( const StateRenderable *s = runTimeCast<const StateRenderable>( it->second.get() ) )
+						{
+							s->render( m_renderer );
+						}
+						else if( const ObjectVector *o = runTimeCast<const ObjectVector>( it->second.get() ) )
+						{
+							for( ObjectVector::MemberContainer::const_iterator it = o->members().begin(), eIt = o->members().end(); it != eIt; it++ )
+							{
+								const StateRenderable *s = runTimeCast<const StateRenderable>( it->get() );
+								if( s )
+								{
+									s->render( m_renderer );
+								}
+							}
+						}
+						else if( const Data *d = runTimeCast<const Data>( it->second.get() ) )
+						{
+							m_renderer->setAttribute( it->first, d );
+						}
+					}
+					s->m_attributes = 0;
+
+					if( m_editMode )
+					{
+						m_renderer->editEnd();
+					}
+
+				}
+
+				// object:
+				if( s->m_object && !m_editMode )
+				{
+					if( const VisibleRenderable *renderable = runTimeCast< const VisibleRenderable >( s->m_object.get() ) )
+					{
+						renderable->render( m_renderer );
+					}
+					s->m_object = 0;
+				}
+			}
+			catch( const std::exception& e )
+			{
+				IECore::msg( IECore::Msg::Error, "InteractiveRender::update", name + ": " + e.what() );
+			}
+
+			return NULL;
+		}
+		
+	private:
+		
+		Renderer *m_renderer;
+		ScenePlug::ScenePath m_previousPath;
+		int m_attrBlockCounter;
+		bool m_editMode;
+};
+
+void InteractiveRender::runPipeline(tbb::pipeline *p)
+{
+	p->run( 2 * tbb::task_scheduler_init::default_num_threads() );
```

Comment:
I've just done some tests on the snowy hillside with cortex 9.0.0-b2 and the latest code in this pull request, and here are some token counts and timings (default_num_threads() was returning 16 on my machine):
```
tokens   startup time
=====================
8        46sec
16       38sec
32       35sec
64       35sec
1000000  30sec
```
I need to test this on a wider range of production data really, but on that scene, 2 * default_num_threads() seems like a reasonable compromise between giving it a chance to do stuff in the background and doing everything at once/keeping loads of unnecessary data around. Maybe I'll just put down a "todo: tune this properly once we've got a load of production examples"